### PR TITLE
Add support for the "Hack" pipeline proposal

### PR DESCRIPTION
**Pull request by js-choi**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | None
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | babel/website#2541
| Any Dependency Changes?  | No
| License                  | MIT

@xixixao and I discussed it a bit more over email, and since he’s dealing with another open-source project, he won’t be able to work on #11600 much for the foreseeable future. So I’ll just open a pull request of my own for the [Hack-pipes proposal](https://github.com/js-choi/proposal-hack-pipes/) under consideration by TC39. CC @TabAtkins, @ljharb, @mAAdhaTTah.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13191"></a>

### CODE REVIEW

**Comment by mAAdhaTTah** [comment]
> Rename "smart" pipe-proposal option to "hack" 

This is going to break backwards compatibility so I don't think we can do this.

**Comment by JLHwung** [comment]
> > Rename "smart" pipe-proposal option to "hack"
> 
> This is going to break backwards compatibility so I don't think we can do this.

It has been reverted in 11a7a49.

**Comment by JLHwung** [comment]
@js-choi You can skip the async function execution tests on Node.js 6 like we did in https://github.com/babel/babel/blob/672a58660f0b15691c44582f1f3fdcdac0fa0d2f/packages/babel-plugin-proposal-do-expressions/test/fixtures/do-expressions/yield/options.json

File: packages/babel-parser/src/plugin-utils.js
Comment (JLHwung) [inline packages/babel-parser/src/plugin-utils.js:101]: Per https://github.com/js-choi/proposal-hack-pipes/issues/3 we should also throw if `topicToken` is `#` and we have `"paritialApplication"` plugin.

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1163]: In Babel 8 we should always use `TopicReference` for both the proposals.

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1165]: This is easily recoverable, but +1 for throwing since we cannot choose between `TopicReference` and `PipelinePrimaryTopicReference`.

File: packages/babel-parser/src/parser/error-message.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/error-message.js:139]: Error codes are now part of the public API, so we should avoid changing them if possible.

File: packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js:21]: This only works in this case because 2.2 doesn't have any side effect, otherwise it's observable.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:398]: Yes, `(a, b) |> #` results in `a`. @ljharb made a good argument for allowing comma expressions for side effects like `console.log` calls, in js-choi/proposal-hack-pipes#1.

File: packages/babel-parser/src/parser/statement.js
Comment (js-choi) [reply packages/babel-parser/src/parser/statement.js:533]: As far as I know, removing `this.withTopicForbiddingContext` is a forwards-compatible change. Is there a particular reason to add it back?

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1203]: My plan had been to remove the smart-mix option in the next major version. Is there a particular reason to keep it?

File: packages/babel-parser/src/parser/error-message.js
Comment (js-choi) [reply packages/babel-parser/src/parser/error-message.js]: This has been addressed in 1c07c3ddc5f97734d02446be10816194dd72aa4c.

**Comment by js-choi** [comment]
All checks now pass in this branch.

**Comment by js-choi** [comment]
An update to the documentation website has been published to babel/website#2541.

Work on [supporting `topicToken: "%"`][another branch] (using `%` instead of `#` as the topic reference) has already been completed in [another branch][]. That branch can be merged into this pull request (its changes are relatively simple), or it could be deferred to a future pull request.

[another branch]: https://github.com/js-choi/babel/tree/proposal-hack-pipes-percent

**Comment by js-choi** [comment]
…I’m not sure why the babel-parser `flow/typecasts/3` test is suddenly not working, only in Node 6. I haven’t modified anything to do with Flow, and the test seems to be working fine in Node 8–14.

**Comment by JLHwung** [comment]
The CI error is not related, see https://github.com/babel/babel/pull/13374 It is likely a bug in the the crankshaft compiler.

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: Oh if the smart proposal is "abandoned" yes, we can just remove it.

**Comment by nicolo-ribaudo** [comment]
> Once this is merged, the next step would probably be updating documentation then adding support for a `"topicToken": "%"` option.

When this PR is ready to be merged (2+ approvals) we won't merge it to the `main` branch until we are ready for a minor release, but I can merge it to a separate branch that so that you can open a PR adding `topicToken: "%"`.

File: packages/babel-parser/src/parser/error-message.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/error-message.js:141]:
```suggestion
    PipelineBodyNoArrow:
```

We should keep the old error message codes now that they are public. After this PR is merged, you could open a new PR to put all the "smart mix" stuff behind a `if (!process.env.BABEL_8_BREAKING)` flag so that they won't be included when working on Babel 8.

File: packages/babel-parser/src/parser/error-message.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/error-message.js:145]:
```suggestion
    PipelineTopicUnused:
```

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1246]: When can this be falsy?

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1260]: Nit (sorry, I'm commenting on the code style):
1. I don't like using `proposalToNodeType[pipeProposal]` much since it can defer to prototype access
2. This arrow-function-just-to-have-throw-expressoin is clever but I don't think I have ever seen it used in our code base

Could you use something like this instead? :pray:
```suggestion
      let nodeType;
      if (pipeProposal === "hack") nodeType = "TopicReference";
      else if (pipeProposal === "smart") nodeType = "PipelinePrimaryTopicReference";
      else throw this.raise(node.start, Errors.PipeTopicRequiresHackPipes);
```

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:2608]: Should this be this?
```suggestion
        this.state.value,
```

Or maybe we could even unify both the arrow and assign cases:
```js
if (this.match(tt.arrow) || this.match(tt.assign)) {
  throw this.raise(this.state.start, Errors.PipeBodyIsTight, this.state.value);
}
```

File: packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js:38]: This doesn't need to be its own function
```suggestion
      if (path.node.operator !== "|>") {
```

File: packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js:47]: Also these two functions can be inlined, since they are used only here.

File: packages/babel-plugin-proposal-pipeline-operator/src/index.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/src/index.js:18]: Not really part of this PR, but we throw an error if `!visitorsPerProposal.hasOwnProperty(options.proposl)`.

File: packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js:19]: Q: Are the changes in this file needed for something? If not, could you revert it? :pray:

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js:3]: Can you also add `input`/`output` fixtures for all these new exec tests?
You can copy the `exec.js` files to `input.js` in the same folders, and just remove the `exec` assertions.

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js:1]: Are these fixtures just moved to somewhere else? I cannot find them :thinking:

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1246]: When someone uses `#` (or `%` in the future) on its own, outside of a pipe expression, and without having turned on the `pipelineOperator` plugin.

I can change the method name to `maybeParseTopicReference` and add a comment to this line, to make that clearer.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:2608]: Unfortunately, I cannot figure out how to detect when an assignment is erroneously used without parentheses (for example, `a |> x = %`, which groups into `(a |> x) = %`), since the parser seems to eagerly consume the `=` before I can detect it at this point. So I’m just getting rid of trying to detect `tt.assign`.

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js:3]: I had thought that `exec.js` tests were preferred to `input`/`output.json`? That’s what [CONTRIBUTING.md](https://github.com/babel/babel/blob/main/CONTRIBUTING.md) says, at least: 

> Generally, there are two kinds of tests for plugins
> * The first is a simple test of the input and output produced by running Babel on some code. We do this by creating an input.js file and an output.js file.  
> * The second and preferred type is a test that actually evaluates the produced code and asserts that certain properties are true or false. We do this by creating an exec.js file.

File: packages/babel-parser/src/parser/error-message.js
Comment (js-choi) [reply packages/babel-parser/src/parser/error-message.js]: This has been addressed in 86c18d3dba0972d2447eff2bd23fe3422d95efcb.

File: packages/babel-parser/src/parser/error-message.js
Comment (js-choi) [reply packages/babel-parser/src/parser/error-message.js]: This has been addressed in 86c18d3dba0972d2447eff2bd23fe3422d95efcb.

File: packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js]: Addressed in 74be75fd1aef92e4ecc0710020710d5122d28275.

File: packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js]: Addressed in 74be75fd1aef92e4ecc0710020710d5122d28275.

File: packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js]: Resolved in faa29eaa4b4add57bbaff99f601f8ca60424ba24.

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js:3]: Oh, wait, so we want *both* `exec.js` *and* `input`/`output.js` files. I can do that.

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js:1247]: Oh a comment would be enough 👍

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1270]: I think there should be one `}` less?

Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1269]: Yes, this was my mistake. My commits got mixed up between my two local branches, so I had to rewrite a bit of history, sorry. I force-pushed a corrected version just now.

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js:1]: Their removal was my mistake. I plan to restore them in a little bit.

File: packages/babel-parser/src/parser/expression.js
Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: Maybe we could add the check in `parseMaybeAssign` (in the same place where we throw about the invalid LHS) but throw the better error message?

Or maybe better, we can check if `this.match(tt.assign)` after https://github.com/js-choi/babel/blob/c9823c517728a383b45387ab1b7205217fe67956/packages/babel-parser/src/parser/expression.js#L531.

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/test/fixtures/smart/nested/input.js:1]: This was addressed in f9a1f1386de3bec4ad5e8a23a4695fe4db8f824d. 

The missing smart-mix tests were also duplicated to test Hack pipes. This uncovered a bug with Hack pipes (pipes within assignment expressions are not working), which I will work on later.

**Comment by js-choi** [comment]
Adding some more tests found a bug, in which Hack pipes within assignment expressions do not transform correctly. I will work on debugging this when I can.

**Comment by js-choi** [comment]
The problem appears to be that `traverse` in https://github.com/babel/babel/blob/6d909afb71d32a24f7ec162b4615fd3be0dbc98d/packages/babel-plugin-proposal-pipeline-operator/src/hackVisitor.js#L30-L33 

…surprisingly does not actually apply the visitor to the path’s root node, but rather to any children that the root node contains. So when `path.get("right")` itself points to a `TopicReference` node (like with `v |> #`), rather than a branch node containing a `TopicReference` node (like with `v |> # + 1`), then the visitor will never be applied to the `TopicReference` node…

File: packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/test/fixtures/hack-hash/basic/exec.js]: Fixed in 2bd98d2b1bc09d590fb7cca04159d0f31a33bfb0.

File: packages/babel-plugin-proposal-pipeline-operator/src/index.js
Comment (js-choi) [reply packages/babel-plugin-proposal-pipeline-operator/src/index.js]: Configuring the plugin without a `proposal` already throws an error: https://github.com/js-choi/babel/blob/11a7a491b57c7c1e2b33cf99a9d244c6eb6e32d6/packages/babel-plugin-syntax-pipeline-operator/src/index.js#L11-L16.

Do we still want to throw an error specifically in `babel-plugin-proposal-pipeline-operator/src/index.js`, too?

I added a test case covering the no-proposal case in 518cf048a215af9f9a872129e325e0affbe86dd4.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1250]: Addressed in fc176687852f984aeddf8814bdf24a58bbce0b47.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js]: `parseMaybeAssign` turned out to be the best place to handle this, especially given the plurality of assignment operators. This has been resolved in cd8a457f77b75618f43dcb260f677e3b52620641.

File: packages/babel-parser/typings/babel-parser.d.ts
Comment (JLHwung) [inline packages/babel-parser/typings/babel-parser.d.ts:162]: Is `topicToken` required in operator plugin options? Maybe `topicToken?`?

**Review by JLHwung — APPROVED** [review]
Awesome work! LGTM with some comments.

File: packages/babel-parser/typings/babel-parser.d.ts
Comment (js-choi) [reply packages/babel-parser/typings/babel-parser.d.ts:162]: `topicToken` is purposefully required (see the documentation in babel/website#2541 and also: https://github.com/babel/babel/blob/11a7a491b57c7c1e2b33cf99a9d244c6eb6e32d6/packages/babel-plugin-syntax-pipeline-operator/src/index.js#L11-L16

`topicToken` will be an option between at least two choices, `#` and `%`, with no particular choice currently considered better than the other. In the future, if TC39 advances Hack pipes, then it will bikeshed on the particular topic token and standardize one of them, and then that token would become the default.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js]: A comment to that line was added in 43dd9afde5e0dc62c42bc5bf9d6e4e2b42f2b10c.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js]: The branches have been rearranged in 43dd9afde5e0dc62c42bc5bf9d6e4e2b42f2b10c, such that errors are checked before constructing a node and consuming the token.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1180]: This flag was removed in 92bcdb633a510d9f83ae2702c17285c7f1eca11c.

File: packages/babel-parser/src/parser/statement.js
Comment (js-choi) [reply packages/babel-parser/src/parser/statement.js:533]: `withTopicForbiddingContext` was restored in 46d9f8ce4170609808219c1bfe15d24313bad30f.

File: packages/babel-parser/typings/babel-parser.d.ts
Comment (js-choi) [reply packages/babel-parser/typings/babel-parser.d.ts:162]: …Oh, wait, I’m being silly. Of course `topicToken` is optional…when `proposal` isn’t `"hack"`.

Comment (js-choi) [reply packages/babel-parser/typings/babel-parser.d.ts]: Fixed in 01ca985505265ec81ba7be05cb3aa22d208f7021.

**Comment by js-choi** [comment]
Not sure why the create-react-app check is failing; unable to reproduce it.

**Comment by js-choi** [comment]
Need to add a check for the conflict between `pipelineOperator: { topicToken: "#" }` and `recordAndtuple: { syntaxType: "hash"}` due to `#[]`.

**Comment by nicolo-ribaudo** [comment]
> I’m not sure why the create-react-app CI check is failing; I am unable to reproduce it.

e2e tests are sometimes flaky; that specific failure seems completely unrelated to this PR.

File: packages/babel-parser/src/parser/expression.js
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js:1325]: I don’t think this will be a problem. The tokenizer consumes the `#[` preferentially to `#`, which means that `maybeParseTopicReference` will never be called.

I added a [test for combining `topicToken: "%"` with hash tuples](https://github.com/js-choi/babel/tree/proposal-hack-pipes-percent/packages/babel-parser/test/fixtures/experimental/pipeline-operator/hack-percent-proposal-hash-tuple) in the [`js-choi:proposal-hack-pipes-percent` branch](https://github.com/js-choi/babel/tree/proposal-hack-pipes-percent/). This test checks that `a |> #[0, %]` correctly parses.

Incidentally, we are [unlikely to choose `?` as a topic token](https://github.com/js-choi/proposal-hack-pipes/issues/2#issuecomment-849179767), so it may not be worth adding support for it to Babel. `%` and `#` are the current frontrunners.

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Do you want me to squash-merge this into a `feat-7.15.0/hack-pipes` branch so that you can open a PR for `%` on top of that?

**Comment by js-choi** [comment]
@nicolo-ribaudo Sounds good. 👍

**Comment by nicolo-ribaudo** [comment]
Merged in https://github.com/babel/babel/tree/feat-7.15.0/hack-pipes! I you'll need that branch to be rebased let me know.

**Pull request merged**
